### PR TITLE
SLVS-655 Rework DisposableNotDisposed to not be based on compilation events & fix FPs/FNs

### PR DESCRIPTION
**Pull request by dbolkensteyn**

### CODE REVIEW

**Comment by dbolkensteyn** [comment]
This rule's execution time went up from 2.2 seconds to 4.3, but is now way more responsive (as in real-time) on large projects.

**Comment by tvsonar** [comment]
This PR fixes FP&FNs, so it would also be good be place to add `HttpClient` to the `TrackedTypes`.

File: src/SonarLint.CSharp/Rules/DisposableNotDisposed.cs [hunk 1]
```diff
@@ -91,89 +91,107 @@ private class NodeAndSymbol
...
         {
-            context.RegisterCompilationStartAction(
-                analysisContext =>
+            context.RegisterSymbolAction(
+                c =>
                 {
-                    var trackedNodesAndSymbols = new HashSet<NodeAndSymbol>();
-                    var excludedSymbols = new HashSet<ISymbol>();
+                    var namedType = (INamedTypeSymbol)c.Symbol;
+                    if (namedType.TypeKind != TypeKind.Class && namedType.TypeKind != TypeKind.Struct || namedType.ContainingType != null)
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/DisposableNotDisposed.cs]: Parentheses should be added to explicitly mark the precedence of `&&` and `||`. Also, we could swap the two sides of the `||`

File: src/SonarLint.CSharp/Rules/DisposableNotDisposed.cs [hunk 2]
```diff
@@ -91,89 +91,107 @@ private class NodeAndSymbol
...
+                    {
+                        return;
+                    }
 
-                    TrackInitializedLocalsAndPrivateFields(analysisContext, trackedNodesAndSymbols);
-                    TrackAssignmentsToLocalsAndPrivateFields(analysisContext, trackedNodesAndSymbols);
+                    var typesDeclarationAndCompilation =
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/DisposableNotDisposed.cs]: misleading variable name, it contains `SyntaxNode`-`SemanticModel` pairs, and not `Compilation`s

File: src/SonarLint.CSharp/Rules/DisposableNotDisposed.cs [hunk 3]
```diff
@@ -182,29 +200,32 @@ private static bool IsLocalOrPrivateField(ISymbol symbol)
...
-            context.RegisterSyntaxNodeAction(
-                c =>
-                {
+            var incovationsAndConditionalAccesses = typeDeclaration
+                .DescendantNodes()
+                .Where(n => n.IsKind(SyntaxKind.InvocationExpression) || n.IsKind(SyntaxKind.ConditionalAccessExpression));
+
+            foreach (SyntaxNode incovationOrConditionalAccess in incovationsAndConditionalAccesses)
+            {
                     SimpleNameSyntax name;
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/DisposableNotDisposed.cs]: too long indentation

File: src/SonarLint.CSharp/Rules/DisposableNotDisposed.cs [hunk 4]
```diff
@@ -219,55 +240,54 @@ private static void ExcludeDisposedAndClosedLocalsAndPrivateFields(CompilationSt
...
+                        excludedSymbols.Add(referencedSymbol);
                     }
-                },
-                SyntaxKind.IdentifierName,
-                SyntaxKind.SimpleMemberAccessExpression);
+                }
+            }
         }
 
         private static bool IsStandaloneExpression(ExpressionSyntax expression)
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/DisposableNotDisposed.cs]: The body if this method still bothers me.

**Comment by tvsonar** [comment]
We have false negatives on arrays, but I don't think we can do a lot about it:
```
private static IDisposable[] field_fs3 = new[] { new FileStream(@"c:\foo.txt", FileMode.Open) };
```
We could check if inside the class there's at least one `field_fs3[X].Dispose()` or `.Close()` with any index or if the `field_fs3` or any `field_fs3[X]` is passed around.

**Comment by tvsonar** [comment]
LGTM

**Pull request merged**